backend/worker/error_scorer.go

```diff
@@ -0,0 +1,283 @@
+		return
+	}
+
+	if len(errorGroups) == 0 {
+		log.WithContext(ctx).WithField("lastErrorObjectId", lastComputedErrorInstanceId).Info("No errors to score")
+		return
+	}
+
+	// TODO(spenny): currently N+1 query
+	for _, errorGroup := range errorGroups {
+		occurranceScore, affectedUsersScore, err := errorScorer.scoreErrorGroup(ctx, errorGroup)
+		_ = errorScorer.redis.SetLastComputedImpactfulErrorObjectId(ctx, errorGroup.LastErrorObjectId)
+		if err != nil {
+			log.WithContext(ctx).WithField("errorGroupId", errorGroup.ID).Error(err)
+			continue
+		}
+
+		totalScore := occurranceScore + affectedUsersScore
+		currentTime := time.Now()
+
+		if err := errorScorer.db.Model(&model.ErrorGroup{Model: model.Model{ID: errorGroup.ID}}).Updates(&model.ErrorGroup{ImpactfulScore: &totalScore, ImpactfulScoreDate: &currentTime}).Error; err != nil {
+			log.WithContext(ctx).WithField("errorGroupId", errorGroup.ID).Error(err)
+			continue
+		}
+
+		log.WithContext(ctx).WithFields(log.Fields{"errorGroupId": errorGroup.ID, "occurranceScore": occurranceScore, "affectedUsersScore": affectedUsersScore}).Info("Scored error group")
+	}
+}
+
+func (errorScorer *ErrorScorer) scoreErrorGroup(ctx context.Context, errorGroup *ErrorGroupWithLastObjectId) (float64, float64, error) {
+	newLimitedDataError := errorGroup.CreatedAt.After(time.Now().Add(-time.Hour * 24 * 7))
+
+	var occuranceScore sql.NullFloat64
+	var affectedUsersScore sql.NullFloat64
+
+	if newLimitedDataError {
+		// number of occurances
+		if err := errorScorer.db.Raw(`
+			WITH most_occurring_errors AS (
+				SELECT error_group_id, COUNT(*) as error_count
+				FROM error_objects
+				WHERE created_at > NOW() - INTERVAL '7 day'
+					AND project_id = ?
+				GROUP BY error_group_id
+				ORDER BY error_count DESC
+				LIMIT 5
+			),
+			current_error AS (
+				SELECT COUNT(*) as current_error_count
+				FROM error_objects
+				WHERE created_at > NOW() - INTERVAL '1 hour'
+					AND error_group_id = ?
+			),
+			trailing_vals AS (
+				SELECT error_objects.error_group_id, extract(day from created_at) AS day, EXTRACT(hour from created_at) AS hour, COUNT(*) AS count
+				FROM error_objects
+				INNER JOIN most_occurring_errors
+					ON most_occurring_errors.error_group_id = error_objects.error_group_id
+				WHERE created_at > NOW() - INTERVAL '7 days'
+				GROUP BY error_objects.error_group_id, day, hour
+				ORDER BY day DESC, hour DESC
+			),
+			stddev AS (
+				SELECT STDDEV_POP(count) FROM trailing_vals
+			),
+			avg AS (
+				SELECT AVG(count) FROM trailing_vals
+			)
+			SELECT (current_error_count - avg) / NULLIF(stddev_pop, 0) AS score
+			FROM trailing_vals, stddev, avg, current_error
+			ORDER BY day DESC, hour DESC
+			LIMIT 1
+		`, errorGroup.ProjectID, errorGroup.ID).Find(&occuranceScore).Error; err != nil {
+			return 0, 0, errors.Wrap(err, "error calculating occurance score")
+		}
+
+		// number of users affected
+		if err := errorScorer.db.Raw(`
+			WITH most_occurring_errors AS (
```

Comment:
This one also uses index scans, but is a little concerning with the execution time (~21 seconds). Maybe we want to use Clickhouse here to improve this query?

```
-[ RECORD 1 ]-------------------------
QUERY PLAN | Limit  (cost=486745.40..486745.40 rows=1 width=72) (actual time=21589.333..21589.348 rows=1 loops=1)
-[ RECORD 2 ]-------------------------
QUERY PLAN |   CTE trailing_vals
-[ RECORD 3 ]-------------------------
QUERY PLAN |     ->  Merge Join  (cost=438823.16..439265.23 rows=458 width=72) (actual time=21212.388..21216.694 rows=103 loops=1)
-[ RECORD 4 ]-------------------------
QUERY PLAN |           Merge Cond: (((EXTRACT(day FROM sessions_2.created_at)) = (EXTRACT(day FROM sessions_3.created_at))) AND ((EXTRACT(hour FROM sessions_2.created_at)) = (EXTRACT(hour FROM sessions_3.created_at))))
-[ RECORD 5 ]-------------------------
QUERY PLAN |           ->  GroupAggregate  (cost=395000.91..395059.79 rows=2141 width=76) (actual time=8839.846..8841.355 rows=103 loops=1)
-[ RECORD 6 ]-------------------------
QUERY PLAN |                 Group Key: (EXTRACT(day FROM sessions_2.created_at)), (EXTRACT(hour FROM sessions_2.created_at)), error_objects_1.error_group_id
-[ RECORD 7 ]-------------------------
QUERY PLAN |                 ->  Sort  (cost=395000.91..395006.26 rows=2141 width=92) (actual time=8839.773..8839.980 rows=3961 loops=1)
-[ RECORD 8 ]-------------------------
QUERY PLAN |                       Sort Key: (EXTRACT(day FROM sessions_2.created_at)) DESC, (EXTRACT(hour FROM sessions_2.created_at)) DESC, error_objects_1.error_group_id
-[ RECORD 9 ]-------------------------
QUERY PLAN |                       Sort Method: quicksort  Memory: 374kB
-[ RECORD 10 ]-------------------------
QUERY PLAN |                       ->  Nested Loop  (cost=48547.68..394882.47 rows=2141 width=92) (actual time=2666.386..8837.537 rows=3961 loops=1)
-[ RECORD 11 ]-------------------------
QUERY PLAN |                             ->  Nested Loop  (cost=48547.10..55162.98 rows=83151 width=8) (actual time=2137.853..2580.666 rows=62751 loops=1)
-[ RECORD 12 ]-------------------------
QUERY PLAN |                                   ->  Limit  (cost=48546.53..48546.54 rows=5 width=12) (actual time=2114.953..2114.962 rows=5 loops=1)
-[ RECORD 13 ]-------------------------
QUERY PLAN |                                         ->  Sort  (cost=48546.53..48575.26 rows=11494 width=12) (actual time=2114.952..2114.958 rows=5 loops=1)
-[ RECORD 14 ]-------------------------
QUERY PLAN |                                               Sort Key: (count(*)) DESC
-[ RECORD 15 ]-------------------------
QUERY PLAN |                                               Sort Method: top-N heapsort  Memory: 25kB
-[ RECORD 16 ]-------------------------
QUERY PLAN |                                               ->  HashAggregate  (cost=48240.68..48355.62 rows=11494 width=12) (actual time=2114.861..2114.926 rows=145 loops=1)
-[ RECORD 17 ]-------------------------
QUERY PLAN |                                                     Group Key: error_objects_2.error_group_id
-[ RECORD 18 ]-------------------------
QUERY PLAN |                                                     Batches: 1  Memory Usage: 417kB
-[ RECORD 19 ]-------------------------
QUERY PLAN |                                                     ->  Index Scan using idx_error_objects_created_at on error_objects error_objects_2  (cost=0.57..48033.94 rows=41348 width=4) (actual time=1.838..2105.692 rows=65521 loops=1)
-[ RECORD 20 ]-------------------------
QUERY PLAN |                                                           Index Cond: (created_at > (now() - '7 days'::interval))
-[ RECORD 21 ]-------------------------
QUERY PLAN |                                                           Filter: (project_id = 1)
-[ RECORD 22 ]-------------------------
QUERY PLAN |                                                           Rows Removed by Filter: 716004
-[ RECORD 23 ]-------------------------
QUERY PLAN |                                   ->  Index Only Scan using idx_error_objects_error_group_id_session_id_v2 on error_objects error_objects_1  (cost=0.57..1156.98 rows=16630 width=8) (actual time=22.356..91.357 rows=12550 loops=5)
-[ RECORD 24 ]-------------------------
QUERY PLAN |                                         Index Cond: (error_group_id = error_objects_2.error_group_id)
-[ RECORD 25 ]-------------------------
QUERY PLAN |                                         Heap Fetches: 57867
-[ RECORD 26 ]-------------------------
QUERY PLAN |                             ->  Memoize  (cost=0.58..6.31 rows=1 width=36) (actual time=0.099..0.099 rows=0 loops=62751)
-[ RECORD 27 ]-------------------------
QUERY PLAN |                                   Cache Key: error_objects_1.session_id
-[ RECORD 28 ]-------------------------
QUERY PLAN |                                   Cache Mode: logical
-[ RECORD 29 ]-------------------------
QUERY PLAN |                                   Hits: 61678  Misses: 1073  Evictions: 0  Overflows: 0  Memory Usage: 83kB
-[ RECORD 30 ]-------------------------
QUERY PLAN |                                   ->  Index Scan using sessions_pkey on sessions sessions_2  (cost=0.57..6.30 rows=1 width=36) (actual time=5.783..5.783 rows=0 loops=1073)
-[ RECORD 31 ]-------------------------
QUERY PLAN |                                         Index Cond: (id = error_objects_1.session_id)
-[ RECORD 32 ]-------------------------
QUERY PLAN |                                         Filter: (created_at > (now() - '7 days'::interval))
-[ RECORD 33 ]-------------------------
QUERY PLAN |                                         Rows Removed by Filter: 1
-[ RECORD 34 ]-------------------------
QUERY PLAN |           ->  GroupAggregate  (cost=43822.25..44035.95 rows=8548 width=72) (actual time=12372.483..12375.238 rows=167 loops=1)
-[ RECORD 35 ]-------------------------
QUERY PLAN |                 Group Key: (EXTRACT(day FROM sessions_3.created_at)), (EXTRACT(hour FROM sessions_3.created_at))
-[ RECORD 36 ]-------------------------
QUERY PLAN |                 ->  Sort  (cost=43822.25..43843.62 rows=8548 width=88) (actual time=12372.450..12372.791 rows=5743 loops=1)
-[ RECORD 37 ]-------------------------
QUERY PLAN |                       Sort Key: (EXTRACT(day FROM sessions_3.created_at)) DESC, (EXTRACT(hour FROM sessions_3.created_at)) DESC
-[ RECORD 38 ]-------------------------
QUERY PLAN |                       Sort Method: quicksort  Memory: 570kB
-[ RECORD 39 ]-------------------------
QUERY PLAN |                       ->  Index Scan using idx_sessions_project_id_include_fingerprint on sessions sessions_3  (cost=0.70..43264.01 rows=8548 width=88) (actual time=2366.389..12364.840 rows=5825 loops=1)
-[ RECORD 40 ]-------------------------
QUERY PLAN |                             Index Cond: ((project_id = 1) AND (created_at > (now() - '7 days'::interval)))
-[ RECORD 41 ]-------------------------
QUERY PLAN |   ->  Sort  (cost=47480.17..47481.32 rows=458 width=72) (actual time=21589.332..21589.339 rows=1 loops=1)
-[ RECORD 42 ]-------------------------
QUERY PLAN |         Sort Key: trailing_vals.day DESC, trailing_vals.hour DESC
-[ RECORD 43 ]-------------------------
QUERY PLAN |         Sort Method: top-N heapsort  Memory: 25kB
-[ RECORD 44 ]-------------------------
QUERY PLAN |         ->  Nested Loop  (cost=47457.16..47477.88 rows=458 width=72) (actual time=21589.273..21589.310 rows=103 loops=1)
-[ RECORD 45 ]-------------------------
QUERY PLAN |               ->  Nested Loop  (cost=47457.16..47457.27 rows=1 width=32) (actual time=21589.266..21589.274 rows=1 loops=1)
-[ RECORD 46 ]-------------------------
QUERY PLAN |                     ->  Nested Loop  (cost=41707.89..41707.97 rows=1 width=24) (actual time=21216.777..21216.783 rows=1 loops=1)
-[ RECORD 47 ]-------------------------
QUERY PLAN |                           ->  Nested Loop  (cost=20.61..20.66 rows=1 width=16) (actual time=21216.752..21216.755 rows=1 loops=1)
-[ RECORD 48 ]-------------------------
QUERY PLAN |                                 ->  Aggregate  (cost=10.31..10.32 rows=1 width=8) (actual time=21216.729..21216.730 rows=1 loops=1)
-[ RECORD 49 ]-------------------------
QUERY PLAN |                                       ->  CTE Scan on trailing_vals trailing_vals_1  (cost=0.00..9.16 rows=458 width=8) (actual time=21212.390..21216.714 rows=103 loops=1)
-[ RECORD 50 ]-------------------------
QUERY PLAN |                                 ->  Aggregate  (cost=10.31..10.32 rows=1 width=8) (actual time=0.020..0.021 rows=1 loops=1)
-[ RECORD 51 ]-------------------------
QUERY PLAN |                                       ->  CTE Scan on trailing_vals trailing_vals_2  (cost=0.00..9.16 rows=458 width=8) (actual time=0.000..0.010 rows=103 loops=1)
-[ RECORD 52 ]-------------------------
QUERY PLAN |                           ->  Aggregate  (cost=41687.27..41687.28 rows=1 width=8) (actual time=0.023..0.025 rows=1 loops=1)
-[ RECORD 53 ]-------------------------
QUERY PLAN |                                 ->  Hash Join  (cost=41212.53..41687.27 rows=1 width=24) (actual time=0.021..0.022 rows=0 loops=1)
-[ RECORD 54 ]-------------------------
QUERY PLAN |                                       Hash Cond: (error_objects.session_id = sessions.id)
-[ RECORD 55 ]-------------------------
QUERY PLAN |                                       ->  Index Only Scan Backward using idx_error_objects_error_group_id_session_id_v2 on error_objects  (cost=0.57..457.18 rows=6908 width=4) (actual time=0.020..0.020 rows=0 loops=1)
-[ RECORD 56 ]-------------------------
QUERY PLAN |                                             Index Cond: (error_group_id = 210345215)
-[ RECORD 57 ]-------------------------
QUERY PLAN |                                             Heap Fetches: 0
-[ RECORD 58 ]-------------------------
QUERY PLAN |                                       ->  Hash  (cost=40520.93..40520.93 rows=55282 width=28) (never executed)
-[ RECORD 59 ]-------------------------
QUERY PLAN |                                             ->  Index Scan using idx_sessions_payload_created_at on sessions  (cost=0.58..40520.93 rows=55282 width=28) (never executed)
-[ RECORD 60 ]-------------------------
QUERY PLAN |                                                   Index Cond: (created_at > (now() - '01:00:00'::interval))
-[ RECORD 61 ]-------------------------
QUERY PLAN |                     ->  Aggregate  (cost=5749.27..5749.28 rows=1 width=8) (actual time=372.451..372.453 rows=1 loops=1)
-[ RECORD 62 ]-------------------------
QUERY PLAN |                           ->  Bitmap Heap Scan on sessions sessions_1  (cost=5548.31..5749.15 rows=50 width=24) (actual time=370.416..372.419 rows=55 loops=1)
-[ RECORD 63 ]-------------------------
QUERY PLAN |                                 Recheck Cond: ((created_at > (now() - '01:00:00'::interval)) AND (project_id = 1))
-[ RECORD 64 ]-------------------------
QUERY PLAN |                                 Heap Blocks: exact=55
-[ RECORD 65 ]-------------------------
QUERY PLAN |                                 ->  BitmapAnd  (cost=5548.31..5548.31 rows=50 width=0) (actual time=369.030..369.031 rows=0 loops=1)
-[ RECORD 66 ]-------------------------
QUERY PLAN |                                       ->  Bitmap Index Scan on idx_sessions_payload_created_at  (cost=0.00..1943.19 rows=55282 width=0) (actual time=112.693..112.693 rows=568351 loops=1)
-[ RECORD 67 ]-------------------------
QUERY PLAN |                                             Index Cond: (created_at > (now() - '01:00:00'::interval))
-[ RECORD 68 ]-------------------------
QUERY PLAN |                                       ->  Bitmap Index Scan on idx_sessions_project_id  (cost=0.00..3604.84 rows=254969 width=0) (actual time=248.231..248.231 rows=292838 loops=1)
-[ RECORD 69 ]-------------------------
QUERY PLAN |                                             Index Cond: (project_id = 1)
-[ RECORD 70 ]-------------------------
QUERY PLAN |               ->  CTE Scan on trailing_vals  (cost=0.00..9.16 rows=458 width=64) (actual time=0.001..0.015 rows=103 loops=1)
-[ RECORD 71 ]-------------------------
QUERY PLAN | Planning Time: 1.440 ms
-[ RECORD 72 ]-------------------------
QUERY PLAN | Execution Time: 21592.739 ms
```